homeassistant/components/ws66i/coordinator.py

```diff
@@ -0,0 +1,66 @@
+"""Coordinator for WS66i."""
+from __future__ import annotations
+
+from datetime import timedelta
+import logging
+
+from pyws66i import WS66i, ZoneStatus
+
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+_LOGGER = logging.getLogger(__name__)
+
+POLL_INTERVAL = timedelta(seconds=30)
+
+
+class Ws66iDataUpdateCoordinator(DataUpdateCoordinator):
+    """DataUpdateCoordinator to gather data for WS66i Zones."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        my_api: WS66i,
+        zones: list[int],
+    ) -> None:
+        """Initialize DataUpdateCoordinator to gather data for specific zones."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name="WS66i",
+            update_interval=POLL_INTERVAL,
+        )
+        self._ws66i = my_api
+        self._zones = zones
+        self._con_broken = False
+
+    def _update_all_zones(self) -> list[ZoneStatus]:
+        """Fetch data for each of the zones."""
+        if self._con_broken:
+            # Try to re-establish a connection
```

Comment:
Can this be moved into the upstream library? Usually we don't want Home Assistant to have to know about he protocol details if it can be avoided (such as the need to re-establish the connection).  Normally we would expect calling something like 'zone_status' would reconnect if needed

Reply:
I do need to update the upstream library to fix a few minor things, and I think you are right about `zone_status`, however I am not sure if it should reconnect. If `zone_status` could reconnect, then the `open` method almost seems pointless. If applications skip calling `open`, then they might not think about calling `close` when they are done. The way I see it, `open` needs to be called before anything can be done and `close` needs to be called when it is done. I do think though that `zone_status` should raise a `ConnectionError` when the connection is lost. This will alert the application to call `open` again to re-establish it. I don't think the application should know to call `close` like I am doing a few lines down, so the upstream library should take care of that.

If you think `zone_status` should reconnect regardless, I have no issue making that change.

Reply:
It seems like it should auto reconnect if it's in an open state, but not if it's in a closed state.

Raising an exception and handling it is ok as well if there is a use case where it would be a bad idea to auto reconnect

Reply:
Done.